### ISSUE TITLE
Race condition, sometimes default namespace is unnecessarily downloaded

### ISSUE THREAD

**Issue by jlcastillo** [isaachinman/next-i18next#724]
##  Bug Report

In the browser, when we instantiate NextI18Next, the constructor may be called before __NEXT_DATA__ is loaded with the namespaces returned by the server via getInitialProps. When that's the case, the I18N store.ns of the client contains an empty list [] during construction, and is internally populated with the default namespace. Since __NEXT_DATA__ was not loaded yet, the default namespace is not available, and NextI18Next sends a download request to the server.

After a while, when the __NEXT_DATA__ is loaded, all translations seem to work well. Hence it seems the only negative effect is the unneeded download of a JSON file, that in our tests seem to affect the LightHouse score.

This issue is easier to reproduce when Chrome network preset is configured as Online. It is easier to get rid of it when the preset is Fast 3G. When a given deploy shows this issue, the race condition can be reproduced in nearly 100% of cases. When a deploy doesn't show it, it's almost impossible to reproduce it. The same code, deployed to 2 machines, may consistently expose the faulty behaviour in one, and never show it in the other.

## To Reproduce

_app.js
```js
import AnalyticsTracker from "../components/AnalyticsTracker"
import App from "next/app"
import { Provider } from "react-redux"
import React from "react"
import { analyticsHostname } from "../config/index"
import { appWithTranslation } from "../utils/i18n"
import { makeStore } from "../store"
import withRedux from "next-redux-wrapper"

class MyApp extends App {

  static async getInitialProps({ Component, ctx }) {
    const pageProps = Component.getInitialProps ? await Component.getInitialProps(ctx) : {}
    const useAnalytics = ctx.req && ctx.req.hostname === analyticsHostname

    return { pageProps, useAnalytics }
  }

  render() {
    const { Component, pageProps, store } = this.props
    return (
      <Provider store={store}>
        <AnalyticsTracker useAnalytics={this.props.useAnalytics} />
        <Component {...pageProps} />
      </Provider>
    )
  }

}

export default withRedux(makeStore)(appWithTranslation(MyApp))
```

index.js:
```js
import Footer from "../components/Footer"
import Head from "next/head"
import HeaderContainer from "../containers/HeaderContainer"
import Jumbotron from "../components/Jumbotron"
import Main from "../components/Main"
import React from "react"
import { sitemapUrlBase } from "../config"
import { withTranslation } from "../utils/i18n"

const index = function({ i18n }) {
  return (
    <div>
      <Head>
        <title>My Title</title>
        <link rel="canonical" href={sitemapUrlBase + i18n.language} />
        <link rel="alternate" hrefLang="es" href={sitemapUrlBase + "es"} />
        <link rel="alternate" hrefLang="en" href={sitemapUrlBase + "en"} />
      </Head>
      <HeaderContainer />
      <Jumbotron />
      <Main />
      <Footer />
    </div>
  )
}

index.getInitialProps = () => {
  return {
    namespacesRequired: ["header", "jumbotron", "landing_services", "landing_reports", "landing_contact", "footer", "default"]
  }
}

export default withTranslation()(index)
```

utils/i18n.js:
```js
const NextI18Next = require("next-i18next").default

const NextI18NextInstance = new NextI18Next({
  browserLanguageDetection: true,
  defaultLanguage: "es",
  fallbackLng: false,
  otherLanguages: ["en"],

  localeSubpaths: {
    en: "en",
    es: "es"
  },

  whitelist: ["es", "en"],
  load: "currentOnly", // we only provide en, es -> no region specific locals like en-US, es-ES

  ns: [
    "footer",
    "graphics",
    "header",
    "jumbotron",
    "landing_contact",
    "landing_predictions",
    "landing_reports",
    "landing_services",
    "login",
    "signup",
    "default"
  ],
  defaultNS: "default",

  ignoreRoutes: ["/_next/", "/static/", "/public/", "/images/", "/api/", "/sitemap", "/robots.txt", "/fonts/"],

  debug: false,//process.env.NODE_ENV !== "production",
  saveMissing: process.env.NODE_ENV !== "production",

  interpolation: {
    escapeValue: false, // not needed for react!!
    formatSeparator: ",",
    format: (value, format) => {
      if (format === "uppercase") return value.toUpperCase()
      return value
    }
  }
})

module.exports = NextI18NextInstance
```

## Expected behavior

The constructor should not try to download any namespaces until __NEXT_DATA__ is loaded.

## Your Environment

Node 14
Chrome: 83
Windows: 10

next: 9.4.2
next-i18next: 4.4.2
i18next: 19.4.4
react: 16.13.0
redux: 4.0.5

**Comment by isaachinman** [comment]
Can you provide a reproducible repo that doesn't include any of this extra setup?

**Comment by jlcastillo** [comment]
It is a race condition. if I remove stuff I'll change the timing, and you won't be able to reproduce it. Even with the same code sometimes I cannot reproduce it. I debugged it, setting breakpoints inside next-i18next, and I realized than when it happens, next-i18next reaches the line in this screenshot with pending = ["es|default"].


<img width="684" alt="next-i18next_bug_1" src="https://user-images.githubusercontent.com/10090510/83041943-548bfc00-a041-11ea-9c6f-15af2f502d3a.png">

Up in the call stack, this is the state of i18n in the function prepareLoading():

<img width="698" alt="next-i18next_bug_2" src="https://user-images.githubusercontent.com/10090510/83042581-18a56680-a042-11ea-8c61-84025a14d46d.png">

Comparing with the cases where it actually works (without unnecessarily downloading default.json), in those cases the store is populated with the translations in __NEXT_DATA__ section.

Can I provide you any other debug info, appart from making a reproducible repo?

**Comment by isaachinman** [comment]
I need to be able to reproduce this race condition in a repository, to determine if this has anything to do with `next-i18next`, or not.

**Comment by isaachinman** [comment]
@jlcastillo Any update?

**Comment by dcodus** [comment]
@isaachinman Can confirm we have observed the exact same issue, which as @jlcastillo pointed out, does have an impact on performance. I believe i can reliable replicate the issue, however in our case it's the exact opposite where by throttling the network (Fast 3G in chrome and Good 3G in Firefox) we always see the default namespace being fetched even thought it's included in the `initialI18nStore` state. Will try and see if we can get this replicated in a repo.

**Comment by isaachinman** [comment]
OK, so the prevailing thought is that this is somehow a race condition on the client side startup? The initial i18n store, including namespaces, is serialised as `initialI18nStore` into `pageProps`, so I'm a bit curious as to how this could be happening.

If anyone can put together a minimal repro, I will debug it! Thanks.

**Comment by isaachinman** [comment]
Hi there, I have just released `next-i18next@8.0.0`, which uses the internationalised routing built into NextJs v10+. This new version of `next-i18next` is quite different, thus making this PR no longer relevant. Thanks!
